RFC-0003-functions-stats.md

```diff
@@ -0,0 +1,373 @@
+  double avgRowSizeAdjustFactor{1.0};
+};
+```
+
+Add a virtual function to `VectorFunctions.h` which a function can override with the function specific 
+statistics propagation override, e.g. 
+
+```cpp
+@@ -109,6 +160,11 @@ class VectorFunction {
+     return false;
+   }
+ 
++  // How this function impacts statistics.
++  virtual std::optional<VectorFunctionStatistics> statsResolver() const {
++    return std::nullopt;
++  }
++
+```
+
+```cpp
+@@ -129,6 +129,10 @@ class UpperLowerTemplateFunction : public exec::VectorFunction {
+   bool propagateStringEncodingFromAllInputs() const override {
+     return true;
+   }
++
++  std::optional<exec::VectorFunctionStatistics> statsResolver() const override {
++    return std::optional{exec::VectorFunctionStatistics{true}};
++  }
+ };
+```
+
+ * JAVA: Provide a new Annotation i.e. `ScalarFunctionStatsCalculator` in java with fields as follows:
+
+```java
+/**
+ * By default, a function is just a “black box” that the database system knows very little about the behavior of.
+ * However, that means that queries using the function may be executed much less efficiently than they could be.
+ * It is possible to supply additional knowledge that helps the planner optimize function calls.
+ * Scalar functions are straight forward to optimize and can have impact on the overall query performance.
+ * Use this annotation to provide information regarding how this function impacts following query statistics.
+ * <p>
+ * A function may take one or more input column or a constant as parameters. Precise stats may depend on the input
+ * parameters. This annotation does not cover all the possible cases.
+ * </p>
+ * <ul>
+ * <li>
+ * 1. Provide adjustment factor for computing `nullFraction` and `avgRowSize`. These adjustment factor will be
+ * applied to input columns on which this function operates on. 1.0 is the default value indicating
+ * Propagate input columns stats as is.
+ * </li>
+ * <li>
+ * 2. How to estimate number of distinct values. Does this function produces an intersection of
+ * Distinct value counts of each `input column param` or a union.
+ * </li>
+ * </ul>
+ */
+@Retention(RUNTIME)
+@Target({METHOD, TYPE})
+public @interface ScalarFunctionStatsCalculator
+{
+  /**
+   * Disable or enable propagate stats behaviour. A function (e.g. upper(Slice)) may not alter any source stats for the
+   * input column passed as an argument. In this case it might be possible to just set propagate stats as true.
+   * In cases where number of input columns are more than one, one has few choices i.e. is the resulting statistics
+   * a union of all columns, Max, Sum or an intersection. To achieve this behaviour please set `statsResolver` field
+   * in conjunction of setting this. For more precise control consider supplying a callback via ... see Phase 3.
+   */
+  boolean propagateStats() default false;
+
+  /**
+   * Possible values: Max, Sum, union and intersect.
+   */
+  String statsResolver() default "Max";
```

Comment:
Will the `statsResolver` apply the same mechanism on all types of stats ? E.g say I want to choose the min of the distinct value but max of the null fraction, will that be possible ?

Reply:
Can you illustrate, with an example, a multi parameter function where it would make sense to use one of these stats resolver strategies

Reply:
Great point, 
Can we switch to:
`distinctValueCountMapper : "x + y" or "max(x, y)"`
This is much more powerful, the value x and y represents parameters' `x` and `y`'s distinct value count, and expression can be solved using : [grammer](https://github.com/prestodb/presto/blob/master/presto-parser/src/main/antlr4/com/facebook/presto/type/TypeCalculation.g4).

e.g.

```java
    @Description("returns index of first occurrence of a substring (or 0 if not found)")
    @ScalarFunction("strpos")
    @LiteralParameters({"x", "y"})
    @SqlType(StandardTypes.BIGINT)
    @ScalarFunctionStatsCalculator(propagateStats = true, avgRowSize = 64, distinctValueCountMapper = "x*y" )
     // There may be repeated values at different Rows, but their orders will be diff
    // as a result ndv will increase geometrically.
    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
    {
        return stringPositionFromStart(string, substring, 1);
    }
```

Reply:
I feel like introducing more annotations/string expressions to parse requires additional complexity that provides little benefit. You could just introduce a java function and write code to compute it directly. It would make compiling the logic for compiling the function metadata quite simple as you have a 1:1 mapping of function-stats calculations

You could implement an annotation which is shorthand for basic propagations (e.g. `@ScalarFunctionStats(propagate = true)` to make it a little easier for authors. My opinion though is that if we're requiring the ability interpret expressions you may as well allow the author to write the code to do it directly.

```java
    @Description("returns index of first occurrence of a substring (or 0 if not found)")
    @ScalarFunction("strpos")
    @SqlType(StandardTypes.BIGINT)
    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
    {
        return stringPositionFromStart(string, substring, 1);
    }

    @ScalarFunctionStatsCalculator
    public static VariableStatisticsEstimate statsPropgation(VariableStatisticsEstimate inputs...)
    {
        // inherits values from arg at position 0 by default
        return VariableStatsEstimate.builder(inputs[0])
                       .setAverageRowSize(64)
                       .setDistinctValuesCount(inputs[0].getDistinctValues() * inputs[1].getDistinctValues())
                       .build();
    }   
```